airflow/models/connection.py

```diff
@@ -43,6 +47,30 @@ def parse_netloc_to_hostname(*args, **kwargs):
     return _parse_netloc_to_hostname(*args, **kwargs)
 
 
+def sanitize_conn_id(conn_id: str | None) -> str | None:
+    """
+    Sanitizes the connection id and allows only specific characters to be within. Namely,
+    it allows alphanumeric characters plus the symbols @,#,$,%,&,!,-,_,. and () from 1
+    and up to 200 consecutive matches.
+
+    The character selection is such that it prevents the injection of javascript or
+    executable bits in order to avoid any awkward behavior in the front-end.
+
+    :param conn_id: The connection id to sanitize.
+    :return: the sanitized string, `None` otherwise.
+    """
+    # check if `conn_id` or our match group is `None`
+    if conn_id is None or (res := re2.match(_RE_SANITIZE_CONN_ID, conn_id)) is None:
+        # comment, so it works...?
+        # log.warning(
+        #     "We failed to match `conn_id` with value: %s to the allowed pattern or it was None",
+        #     conn_id,
+        # )
```

Comment:
Left over code?

Reply:
Yes, to indicate where my logging was as per my reply above. Happy to remove it completely if you guys are OK with no logging.

Reply:
If possible I feel a warning would be worthwhile in the CLI and UI, but probably not here.

Also what would happen if the user adds a connection from either CLI or web UI and the ID does not sanitise? Does it fail with a message, or is a connection created silently with a null connection ID, or otherwise?

Reply:
It will throw an exception because of the non-null requirement imposed by the SQLAlchemy. This is because if we do not match, None is returned and thus will trigger this.

Reply:
I would recommend adding some guard in the web UI and CLI so the user gets a softer landing than seeing an exception. It would be a good idea to add validation to the web form so the user cannot submit an invalid conn ID in the first place, for example.

Reply:
That's a larger scope of work I think, this PR tries to sanitize the conn_id at the class level which is the ground truth for all.

Reply:
I would say this PR is arguably a degradation in user experience in its current form. A better approach would be to introduce the UX validation first.

Reply:
Perhaps others could chip in on that? I am not sure I share the same view on the matter.

Reply:
I think it's a very good suggestion. And it should be easy to do

UI:
FAB already has ways of adding custom validation for fields, so we should be able to have 1-1 validation on the UI. The benefit of it is that you won't even be able to submit a bad connection_id value. It's a much better experience and standard practice. 

API:

It should be easy to extend our Open API with data/pattern to valide conn_id at the client side https://swagger.io/docs/specification/data-models/data-types/#pattern

The CLI is less needed -  I think becuase it does not relaly matter, you will either get exception raised in the "DB" code and propagated to the client or some pre-validation in the client. It's the matter of nicely displaying the message if the exception happens.

Reply:
Thanks for your comments, I think given that the connection class is the ground truth for everything the original spirit of this PR was to sanitise the `conn_id` as described in the original issue. This was meant as a good starter issue that would be easily solved in a day or so. However, given how much back and forth we had it seems this is more involved, especially if we need to touch more parts of the project that I am unfamiliar with (e.g. UI code and API). That's why I said to do the said suggested additional changes might be a larger piece of work than I originally thought it would be.

I have pushed the code changes to remove out the commented code so the tests pass but I am not sure if I have time to do the UI and API at the moment.

Reply:
Well. It happens. In software development you never know the end of journey until you hit `merge` button. I would encourage you to continue on that with a slower pace and learn those other things as you go. This is how things work in open-source, the journey and learning is often as important as the final result, also the final result whch is "complete" is better than "half-done" - even if takes way longer - especially for those kind of issues that are not the highest priority.

BTW. The tests do not pass yet.